### ISSUE TITLE
VectorFieldModule, TensorFieldModule, DiffFormModule: Add methods tensor_product, tensor_power, update category of TensorFieldModule

### ISSUE THREAD

**Issue by mkoeppe** [sagemath/sage#34589]
<div id="comment:0"></div>

... as previously done for `FiniteRankFreeModule` etc.

We introduce abstract base classes `ReflexiveModule_*` to unify the implementations of `VectorFieldModule`... and `FiniteRankFreeModule`/`VectorFieldFreeModule`...

Follow-up: #34621 Method `dual_pairing` for modules in `sage.tensor`

CC:  @egourgoulhon @tscrim @mjungmath

Component: **manifolds**

Author: **Matthias Koeppe**

Branch/Commit: **[`74601e6`](https://github.com/sagemath/sagetrac-mirror/commit/74601e6a3a42413400b2718ad2fed987028a30ca)**

Reviewer: **Eric Gourgoulhon**

_Issue created by migration from https://trac.sagemath.org/ticket/34589_

**Comment by mkoeppe** [comment]
Dependencies: **#34448**

**Comment by mkoeppe** [comment]
Branch: **[u/mkoeppe/vectorfieldmodule_tensor_product](https://github.com/sagemath/sagetrac-mirror/tree/u/mkoeppe/vectorfieldmodule_tensor_product)**

**Comment by mkoeppe** [comment]
Commit: **[`31b17b5`](https://github.com/sagemath/sagetrac-mirror/commit/31b17b54e661ab696ca5fda4ef8a3fecb8aee0be)**

**Comment by mkoeppe** [comment]
<div id="comment:3" align="right">comment:3</div>

Probably best to create a class `VectorFieldModule_abstract`

---
Last 10 new commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/620df160109f99781c2b6c72ec596a176de5d2c5"><code>620df16</code></a></td><td><code>src/sage/tensor/modules/tensor_free_submodule.py: Remove tensor_factors, add doctest for construction</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/4c9be6b79b6fa8f79edf3f2298c924d0e185f1fd"><code>4c9be6b</code></a></td><td><code>src/sage/manifolds/differentiable/diff_form_module.py: doc: add that the case p=1 of M parallelizable is implemented via VectorFieldDualFreeModule</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/06710c12cc1d73cafe81e087742a514808985f86"><code>06710c1</code></a></td><td><code>src/sage/manifolds/differentiable/diff_form_module.py, vectorfield_module.py: Update copyright using 'git blame -w --date=format:%Y FILE.py | sort -k2', add to AUTHORS</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/bb614c17c7c9d396be1376b526c5eb52c39cba7d"><code>bb614c1</code></a></td><td><code>Merge #34501</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/4a4a29d6cd7d49d4ea9c1720566ce50ba8f4607c"><code>4a4a29d</code></a></td><td><code>Merge tag '9.7' into t/34448/put_tensor_modules_of_finiterankfreemodule_in_modules___tensorproducts__</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/1f4fb90a273b81a7ea9cc14f97c7842eacd022a8"><code>1f4fb90</code></a></td><td><code>Merge tag '9.8.beta0' into t/34448/put_tensor_modules_of_finiterankfreemodule_in_modules___tensorproducts__</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7c0495988c7962163abf3c11ad200add7feb05cc"><code>7c04959</code></a></td><td><code>src/sage/manifolds/differentiable/diff_form_module.py: Fix typo in docstring</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/fc31cbf584a3de4ffa7005c33a4b8ef4a98d09ca"><code>fc31cbf</code></a></td><td><code>Merge #34501</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/f0def924fea5ab8caaeccd0f89d21d455004a6aa"><code>f0def92</code></a></td><td><code>src/sage/tensor/modules/tensor_free_module.py: Remove redundant method</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/31b17b54e661ab696ca5fda4ef8a3fecb8aee0be"><code>31b17b5</code></a></td><td><code>WIP</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:4"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/e10f9cf17bd06784805914b63b7ae0adc4423c96"><code>e10f9cf</code></a></td><td><code>src/sage/manifolds/differentiable/vectorfield_module.py: Revert change to VectorFieldModule.tensor (not needed yet)</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/ff00c1143e73e0e2163096198266df792737e657"><code>ff00c11</code></a></td><td><code>VectorFieldModule: Add tensor_product, tensor_power</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/8d753dcaf07b85d423661f0128bd6d1fff7b71ee"><code>8d753dc</code></a></td><td><code>src/sage/manifolds/differentiable/vectorfield_module.py (VectorFieldModule_abstract): New</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`31b17b5`](https://github.com/sagemath/sagetrac-mirror/commit/31b17b54e661ab696ca5fda4ef8a3fecb8aee0be)** to **[`8d753dc`](https://github.com/sagemath/sagetrac-mirror/commit/8d753dcaf07b85d423661f0128bd6d1fff7b71ee)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`8d753dc`](https://github.com/sagemath/sagetrac-mirror/commit/8d753dcaf07b85d423661f0128bd6d1fff7b71ee)** to **[`6bada9f`](https://github.com/sagemath/sagetrac-mirror/commit/6bada9f80ec73ba15d326858295b2ba0f34e457a)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:6"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/2830ebc72e79932665207cb9ec4fea6d0dd95769"><code>2830ebc</code></a></td><td><code>TensorFreeModule.tensor_factors: Go through methods tensor_type, base_module</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/6bada9f80ec73ba15d326858295b2ba0f34e457a"><code>6bada9f</code></a></td><td><code>TensorFieldModule.tensor_factors: New</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:7"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/347ffc06ff2b3ad500ac1a76480a4e626606c07c"><code>347ffc0</code></a></td><td><code>TensorFieldModule: Put it in the category Modules.TensorProducts</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`6bada9f`](https://github.com/sagemath/sagetrac-mirror/commit/6bada9f80ec73ba15d326858295b2ba0f34e457a)** to **[`347ffc0`](https://github.com/sagemath/sagetrac-mirror/commit/347ffc06ff2b3ad500ac1a76480a4e626606c07c)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`347ffc0`](https://github.com/sagemath/sagetrac-mirror/commit/347ffc06ff2b3ad500ac1a76480a4e626606c07c)** to **[`7194893`](https://github.com/sagemath/sagetrac-mirror/commit/7194893cd9953bdff6c4ce1c28f8d1163a29d3d5)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:9"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7194893cd9953bdff6c4ce1c28f8d1163a29d3d5"><code>7194893</code></a></td><td><code>src/sage/manifolds/differentiable/tensorfield_module.py: Update copyright according to git blame -w --date=format:%Y FILE | sort -k2</code></td></tr></table>

**Comment by mkoeppe** [comment]
Author: **Matthias Koeppe**

**Comment by mkoeppe** [comment]
<div id="comment:11" align="right">comment:11</div>

This new base class is of course not very pretty

```
+class VectorFieldModule_abstract(UniqueRepresentation, Parent):
+    r"""
+    Abstract base class for modules of vector fields.
+    """
+
+    tensor_power = FiniteRankFreeModule_abstract.tensor_power
+
+    tensor_product = FiniteRankFreeModule_abstract.tensor_product
+
+    tensor = FiniteRankFreeModule_abstract.tensor
```

It could be better to have a common base class for `VectorFieldModule_abstract` and `FiniteRankFreeModule_abstract`. 

Perhaps that could be called `ModuleWithDual_abstract`? (see also #13372)

**Comment by mkoeppe** [comment]
<div id="comment:12" align="right">comment:12</div>

Algebra question - do we know / do we assume that a `VectorFieldModule` is reflexive?

**Comment by mkoeppe** [comment]
<div id="comment:13" align="right">comment:13</div>

(in this case, we could call it `ReflexiveModule_abstract` - for such modules the pair `tensor_type` makes sense.)

**Comment by egourgoulhon** [comment]
<div id="comment:14" align="right">comment:14</div>

Replying to [Matthias Köppe](#comment%3A12):
> Algebra question - do we know / do we assume that a `VectorFieldModule` is reflexive?

I would say that the vector field modules are indeed reflexive for the cases we can treat with Sage, namely those for which the manifold M can be split into a finite number of parallelizable manifolds N_i, i.e. there is a finite open cover of M, (N_i)_{1\leq i \leq p} (p finite), such that for each i, N_i is parallelizable. Let then consider an element f of `X(M)**`, where `X(M)` is the C<sup>oo</sup>(M)-module of vector fields on M. For a in `X(M)*`, f(a) can be reduced to f_i(a_i) on each N_i. Since N_i is parallelizable, X(N_i) is a free module of finite rank and thus reflexive. There exists then a unique v_i in X(N_i) such that f_i(a_i) = a_i(v_i). Then the vector field v in X(M) defined as being equal to v_i on each N_i obeys f(a) = a(v), which shows that the canonical map `X(M) --> X(M)**` is bijective. This is a first quick thought and this should be double checked...

**Comment by egourgoulhon** [comment]
<div id="comment:15" align="right">comment:15</div>

Replying to [Matthias Köppe](#comment%3A13):
> (in this case, we could call it `ReflexiveModule_abstract` - for such modules the pair `tensor_type` makes sense.)

Indeed, if we confirm that the vector field modules are reflexive.

**Comment by mkoeppe** [comment]
<div id="comment:16" align="right">comment:16</div>

Replying to [Eric Gourgoulhon](#comment%3A14):
> I would say that the vector field modules are indeed reflexive for the cases we can treat with Sage, namely those for which the manifold M can be split into a finite number of parallelizable manifolds N_i, i.e. there is a finite open cover of M, (N_i)_{1\leq i \leq p} (p finite), such that for each i, N_i is parallelizable.

Is this splitting into parallelizable manifolds explicitly done somewhere in the code?

I guess a subtlety is that in sage.manifolds a manifold M is thought to be immutable but new submanifolds on it can be declared imperatively. At any point, of course only finitely many submanifolds have been declared in M as a Python object, but it's not clear that M as a mathematical object is really assumed to be covered by only a finite number of parallelizable submanifolds...

**Comment by egourgoulhon** [comment]
<div id="comment:17" align="right">comment:17</div>

Replying to [Matthias Köppe](#comment%3A16):
> Replying to [Eric Gourgoulhon](#comment%3A14):
> > I would say that the vector field modules are indeed reflexive for the cases we can treat with Sage, namely those for which the manifold M can be split into a finite number of parallelizable manifolds N_i, i.e. there is a finite open cover of M, (N_i)_{1\leq i \leq p} (p finite), such that for each i, N_i is parallelizable.
> 
> 
> Is this splitting into parallelizable manifolds explicitly done somewhere in the code?

Not explicitly, but `DifferentiableManifold`'s have the attribute `_parrelizable_parts`; for instance:

```
sage: M = manifolds.Sphere(2, coordinates='stereographic')
sage: M._parallelizable_parts
{Open subset S^2-{NP,SP} of the 2-sphere S^2 of radius 1 smoothly embedded in the Euclidean space E^3,
 Open subset S^2-{NP} of the 2-sphere S^2 of radius 1 smoothly embedded in the Euclidean space E^3,
 Open subset S^2-{SP} of the 2-sphere S^2 of radius 1 smoothly embedded in the Euclidean space E^3}
sage: M._open_covers
[[2-sphere S^2 of radius 1 smoothly embedded in the Euclidean space E^3],
 [Open subset S^2-{NP} of the 2-sphere S^2 of radius 1 smoothly embedded in the Euclidean space E^3,
  Open subset S^2-{SP} of the 2-sphere S^2 of radius 1 smoothly embedded in the Euclidean space E^3]]
```
Actually the parallelizable parts are declared on the fly, in the same spirit as the charts and other structures on the manifod. Even the parallelizable aspect of the manifold can appear quite late after the manifold instanciation. For example, in this [notebook about S^1](https://nbviewer.org/github/sagemanifolds/SageManifolds/blob/master/Notebooks/SM_sphere_S1.ipynb), the manifold is (implicitly) declared parallelizable only in cell no. 21, by declaring a vector frame on it. This action creates the object X(M) as a `VectorFieldFreeModule`. Previously, only the modules X(A) and X(B) had been instanciated.  However, one could imagine to add the keyword argument `parallelizable` to the manifold constructor (with default being `False`).
> 
> I guess a subtlety is that in sage.manifolds a manifold M is thought to be immutable but new submanifolds on it can be declared imperatively. At any point, of course only finitely many submanifolds have been declared in M as a Python object, but it's not clear that M as a mathematical object is really assumed to be covered by only a finite number of parallelizable submanifolds...

You're right, this should be stated in the manifold documentation. The whole idea is described in Sec. 3.2 and 3.3 of this [document](https://ccirm.centre-mersenne.org/item/CCIRM_2018__6_1_A1_0/).
Note that the finite number of parallelizable pieces is not a very severe limitation. For instance this holds for any compact manifold. Actually I don't know any manifold that is not of this type, not that I doubt that some examples can be constructed...

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:18"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/d5e715e7f167b9cf928c7c1455dfdf109e4063e8"><code>d5e715e</code></a></td><td><code>sage.tensor.modules.reflexive_module: New, refactor finite rank free modules and vector field modules through the new classes</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`7194893`](https://github.com/sagemath/sagetrac-mirror/commit/7194893cd9953bdff6c4ce1c28f8d1163a29d3d5)** to **[`d5e715e`](https://github.com/sagemath/sagetrac-mirror/commit/d5e715e7f167b9cf928c7c1455dfdf109e4063e8)**

**Comment by mkoeppe** [comment]
<div id="comment:19" align="right">comment:19</div>

Here's a draft of refactoring along these lines. (Of course src/sage/tensor/modules/reflexive_module.py would still need more documentation.)

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`d5e715e`](https://github.com/sagemath/sagetrac-mirror/commit/d5e715e7f167b9cf928c7c1455dfdf109e4063e8)** to **[`e17b30d`](https://github.com/sagemath/sagetrac-mirror/commit/e17b30d6dd76afe40b9a349f05410d8312fe25ed)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:20"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/e7959c853b12082032d1e972e57222df56b190f0"><code>e7959c8</code></a></td><td><code>Merge tag '9.8.beta1' into t/34589/vectorfieldmodule_tensor_product</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/e17b30d6dd76afe40b9a349f05410d8312fe25ed"><code>e17b30d</code></a></td><td><code>src/sage/tensor/modules/reflexive_module.py: Make all classes subclasses of ..._abstract</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`e17b30d`](https://github.com/sagemath/sagetrac-mirror/commit/e17b30d6dd76afe40b9a349f05410d8312fe25ed)** to **[`ca8b730`](https://github.com/sagemath/sagetrac-mirror/commit/ca8b730c4c0d5c9bf1010a80239030d50788d31f)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:21"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/ca8b730c4c0d5c9bf1010a80239030d50788d31f"><code>ca8b730</code></a></td><td><code>src/sage/tensor/modules/reflexive_module.py: Add documentation</code></td></tr></table>

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -1 +1,5 @@
 ... as previously done for `FiniteRankFreeModule` etc.
+
+We introduce abstract base classes `ReflexiveModule_*` to unify the implementations of `VectorFieldModule`... and `FiniteRankFreeModule`/`VectorFieldFreeModule`...
+
+Follow-up: #34621 Method `dual_pairing` for modules in `sage.tensor`
``````

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`ca8b730`](https://github.com/sagemath/sagetrac-mirror/commit/ca8b730c4c0d5c9bf1010a80239030d50788d31f)** to **[`b7d85cc`](https://github.com/sagemath/sagetrac-mirror/commit/b7d85ccaa9d543e33f0b0176af9cb37ee7b3ea27)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:23"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/b7d85ccaa9d543e33f0b0176af9cb37ee7b3ea27"><code>b7d85cc</code></a></td><td><code>src/sage/tensor/modules/reflexive_module.py: Add more doctests</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:24"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/eb826057f68c761d0cd22a186334e8037670e873"><code>eb82605</code></a></td><td><code>src/sage/tensor/modules/reflexive_module.py: Add doctests for full coverage</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`b7d85cc`](https://github.com/sagemath/sagetrac-mirror/commit/b7d85ccaa9d543e33f0b0176af9cb37ee7b3ea27)** to **[`eb82605`](https://github.com/sagemath/sagetrac-mirror/commit/eb826057f68c761d0cd22a186334e8037670e873)**

**Comment by egourgoulhon** [comment]
<div id="comment:26" align="right">comment:26</div>

Looks nice! Thanks.

An entry for the new file `reflexive_module.py` is missing in `src/doc/en/reference/tensor_free_modules/`, so that the documentation is not generated. I would suggest to add it in `src/doc/en/reference/tensor_free_modules/tensors.rst`. Once this is done,

```
sage -docbuild reference/tensor_free_modules html
```
reveals that some `$` signs should be replaced by backticks:

```diff
-    the `R`-module of linear functionals $\phi:\ M \longrightarrow R$.
+    the `R`-module of linear functionals `\phi:\ M \longrightarrow R`.
```

**Comment by egourgoulhon** [comment]
<div id="comment:27" align="right">comment:27</div>

Travis, Michael, if you have time (not sure this is the case at the moment...), could you comment on the argument sketched in [comment:14](#comment%3A14) ?

**Comment by tscrim** [comment]
<div id="comment:28" align="right">comment:28</div>

I think it is good. I do not quite see how it might break down for more obtuse finite-dimensional examples, but the question is should it? I think no, but I typically think in terms of "nice" objects.

**Comment by egourgoulhon** [comment]
<div id="comment:29" align="right">comment:29</div>

Replying to [Travis Scrimshaw](#comment%3A28):
> I think it is good. I do not quite see how it might break down for more obtuse finite-dimensional examples, but the question is should it? I think no, but I typically think in terms of "nice" objects.

Thank you Travis!

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`eb82605`](https://github.com/sagemath/sagetrac-mirror/commit/eb826057f68c761d0cd22a186334e8037670e873)** to **[`2a7b445`](https://github.com/sagemath/sagetrac-mirror/commit/2a7b445e33f01d047d4834af16749570bee9c892)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:30"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/2a7b445e33f01d047d4834af16749570bee9c892"><code>2a7b445</code></a></td><td><code>src/doc/en/reference/tensor_free_modules/tensors.rst: Add reflexive_module</code></td></tr></table>

**Comment by egourgoulhon** [comment]
<div id="comment:31" align="right">comment:31</div>

I have two questions regarding the inheritance structure:

1/ There is a disymmetry in the MRO between `FiniteRankFreeModule`, `FiniteDualFreeModule` and `TensorFreeModule`: for the first two classes, the inheritance order is `(ReflexiveModule_(base or dual), FiniteRankFreeModule_abstract)`,  while for the third class it is the opposite order: `(FiniteRankFreeModule_abstract, ReflexiveModule_tensor)`. Is this intended?

2/ The class `VectorFieldModule_abstract` has no method at all. Is such a class necessary? The only use I can see is that classes that inherit from it inherit from the pair `(UniqueRepresentation, ReflexiveModule_abstract)`, but it would be more clear to make this explicit in those classes, instead of using `VectorFieldModule_abstract` as a proxy for `(UniqueRepresentation, ReflexiveModule_abstract)`. Also, the name `VectorFieldModule_abstract` could be misleading since for instance `DiffFormModule` inherits from it.

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`2a7b445`](https://github.com/sagemath/sagetrac-mirror/commit/2a7b445e33f01d047d4834af16749570bee9c892)** to **[`05daeee`](https://github.com/sagemath/sagetrac-mirror/commit/05daeeec7c6b96f5bf323b3688676c96ab435398)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:32"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/05daeeec7c6b96f5bf323b3688676c96ab435398"><code>05daeee</code></a></td><td><code>src/sage/tensor/modules/tensor_free_module.py (TensorFreeModule): Make MRO consistent with FiniteRankFreeModule</code></td></tr></table>

**Comment by mkoeppe** [comment]
<div id="comment:33" align="right">comment:33</div>

1) Thanks for catching this, fixed.

**Comment by mkoeppe** [comment]
<div id="comment:34" align="right">comment:34</div>

2) I guess I was planning to move some methods from `VectorFieldModule` to `VectorField_abstract`. Candidates would be `domain`, `ambient_domain`, and perhaps some others. But I agree that the name `VectorFieldModule_abstract` is misleading... Would you have a suggestion for a better name? It also doesn't have to happen on this ticket -- I can just back it out.

**Comment by egourgoulhon** [comment]
<div id="comment:35" align="right">comment:35</div>

Replying to [Matthias Köppe](#comment%3A34):
> 2) I guess I was planning to move some methods from `VectorFieldModule` to `VectorField_abstract`. Candidates would be `domain`, `ambient_domain`, and perhaps some others. But I agree that the name `VectorFieldModule_abstract` is misleading... Would you have a suggestion for a better name? 

Maybe `FieldModule_abstract`?

> It also doesn't have to happen on this ticket -- I can just back it out.

OK.

**Comment by egourgoulhon** [comment]
<div id="comment:36" align="right">comment:36</div>

May I ask what do we choose? Renaming `VectorFieldModule_abstract` to `FieldModule_abstract` (or something similar) or backing it out for this ticket?

**Comment by mkoeppe** [comment]
<div id="comment:37" align="right">comment:37</div>

I'll back it out

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`05daeee`](https://github.com/sagemath/sagetrac-mirror/commit/05daeeec7c6b96f5bf323b3688676c96ab435398)** to **[`ebc2f27`](https://github.com/sagemath/sagetrac-mirror/commit/ebc2f2782f66fc3bda57035413cebbb1f002303e)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:38"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. Last 10 new commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/610545e6e2f83b6200675fd708cfb47b280413cb"><code>610545e</code></a></td><td><code>TensorFreeModule.tensor_factors: Go through methods tensor_type, base_module</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/a6ba1011bb5d8601ed0acf1dbc845297042ea1a1"><code>a6ba101</code></a></td><td><code>TensorFieldModule.tensor_factors: New</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/8e0bf363b0ffab0336e4218fa986fac3bcef4b48"><code>8e0bf36</code></a></td><td><code>TensorFieldModule: Put it in the category Modules.TensorProducts</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/20bd0d3f7ced5dc76e1b9032903254cf9aeb6236"><code>20bd0d3</code></a></td><td><code>src/sage/manifolds/differentiable/tensorfield_module.py: Update copyright according to git blame -w --date=format:%Y FILE | sort -k2</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/38e6b3e414c15dde36ba3e7957539e7727d8a14d"><code>38e6b3e</code></a></td><td><code>sage.tensor.modules.reflexive_module: New, refactor finite rank free modules and vector field modules through the new classes</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/04a4daa81badfde762256b27734b423ca3763449"><code>04a4daa</code></a></td><td><code>src/sage/tensor/modules/reflexive_module.py: Make all classes subclasses of ..._abstract</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/53e944a6a330ab81ea77fceab7d05d6552545a08"><code>53e944a</code></a></td><td><code>src/sage/tensor/modules/reflexive_module.py: Add documentation</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/f107a775fd44066f3abc931c01257f3b13a53f2b"><code>f107a77</code></a></td><td><code>src/doc/en/reference/tensor_free_modules/tensors.rst: Add reflexive_module</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/d2228bd2e747c68fecb69617b49d4cad988682e1"><code>d2228bd</code></a></td><td><code>src/sage/tensor/modules/tensor_free_module.py (TensorFreeModule): Make MRO consistent with FiniteRankFreeModule</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/ebc2f2782f66fc3bda57035413cebbb1f002303e"><code>ebc2f27</code></a></td><td><code>WIP get rid of VectorFieldModule_abstract</code></td></tr></table>

**Comment by mkoeppe** [comment]
Changed dependencies from **#34448** to none

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`ebc2f27`](https://github.com/sagemath/sagetrac-mirror/commit/ebc2f2782f66fc3bda57035413cebbb1f002303e)** to **[`8a2866b`](https://github.com/sagemath/sagetrac-mirror/commit/8a2866bf17cf8119b719741c0b9e0574fee58a04)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:40"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/8a2866bf17cf8119b719741c0b9e0574fee58a04"><code>8a2866b</code></a></td><td><code>get rid of VectorFieldModule_abstract</code></td></tr></table>

**Comment by egourgoulhon** [comment]
<div id="comment:42" align="right">comment:42</div>

Replying to [Matthias Köppe](#comment%3A37):
> I'll back it out

Thanks!

One minor last thing: as revealed by the patchbot, `ReflexiveModule_abstract` is imported by unused in `vectorfield_module.py`.

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`8a2866b`](https://github.com/sagemath/sagetrac-mirror/commit/8a2866bf17cf8119b719741c0b9e0574fee58a04)** to **[`74601e6`](https://github.com/sagemath/sagetrac-mirror/commit/74601e6a3a42413400b2718ad2fed987028a30ca)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:43"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/74601e6a3a42413400b2718ad2fed987028a30ca"><code>74601e6</code></a></td><td><code>Remove unused import</code></td></tr></table>

**Comment by egourgoulhon** [comment]
Reviewer: **Eric Gourgoulhon**

**Comment by egourgoulhon** [comment]
<div id="comment:44" align="right">comment:44</div>

Thanks!

**Comment by mkoeppe** [comment]
<div id="comment:45" align="right">comment:45</div>

Thank you!

**Comment by vbraun** [comment]
Changed branch from **[u/mkoeppe/vectorfieldmodule_tensor_product](https://github.com/sagemath/sagetrac-mirror/tree/u/mkoeppe/vectorfieldmodule_tensor_product)** to **[`74601e6`](https://github.com/sagemath/sagetrac-mirror/commit/74601e6a3a42413400b2718ad2fed987028a30ca)**
